lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1258,11 +1237,109 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
     typeCheckStmt(newBody);
     S->setBody(newBody);
 
-    // Check all the catch clauses independently.
-    for (auto clause : S->getCatches()) {
-      checkCatchStmt(clause);
+    SmallVector<VarDecl *, 8> scratchMemory1;
+    SmallVector<VarDecl *, 8> scratchMemory2;
+
+    auto clauses = S->getCatches();
+    CaseStmt *previousBlock = nullptr;
+    for (auto i = clauses.begin(), e = clauses.end(); i != e; ++i) {
```

Comment:
This code looks nearly identical to what's in `visitSwitchStmt`. Can they be unified?

Reply:
I think they can be if I refactor fallthrough handling a little. I'll look into that asap